### PR TITLE
feat(hypothesis): Added hypothesis as an option 

### PR DESCRIPTION
**Pull request by ayeankit**
## Pull Request description

Issue #23 

## How to test these changes

* ```...```

## Pull Request checklists

This PR is a:
- [ ] bug-fix
- [ ] new feature
- [ ] maintenance

About this PR:
- [ ] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [ ] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:
- [ ] I have reviewed the changes and it contains no misspelling.
- [ ] The code is well commented, especially in the parts that contain more complexity.
- [ ] New and old tests passed locally.

## Reviewer's checklist

Copy and paste this template for your review's note:

```
## Reviewer's Checklist

- [ ] I managed to reproduce the problem locally from the `main` branch
- [ ] I managed to test the new changes locally
- [ ] I confirm that the issues mentioned were fixed/resolved .
```

### CODE REVIEW

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.project_slug}}.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 """Tests for `{{ cookiecutter.project_slug }}` package."""
+{%- if cookiecutter.use_pytest == "yes" -%}
```
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.project_slug}}.py:2]: could you revisit this file again?

keep in mind that it would be possible to have both pytest and hypothesis.
also please, check the result locally to see what is the final result for this file for all the 3 possibilities (pytest, hypothesis, pytest + hypothesis).

also pay attention on the jinja2 modifications %-/-%, %+/+% or just % .. and the combinations among them

File: src/scicookie/{{cookiecutter.project_slug}}/pyproject.toml [hunk 2]
```diff
@@ -51,6 +51,9 @@ pytest = "^7.3.2"
 pytest-cov = "^4.1.0"
 {% endif %}
 {%- endif -%}{#- end of use_pytest -#}
+{%- if cookiecutter.use_hypothesis == "yes" -%}
+hypothesis = "^6.0"
+{%- endif -%}
```
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/pyproject.toml:56]:
```suggestion
{% endif -%}
```

Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/pyproject.toml:56]: you can also use this to play with templates:

http://jinja.quantprogramming.com/

**Comment by ayeankit** [comment]
@xmnlab and @Saransh-cpp  I do not know why its failing the main/test (pull_request). Can you review it?

**Comment by xmnlab** [comment]
> @xmnlab and @Saransh-cpp I do not know why its failing the main/test (pull_request). Can you review it?

all the files need to end with one empty line, no more, no less.
and this PR has 2 files that that problem: https://github.com/osl-incubator/scicookie/actions/runs/5328022194/jobs/9652160700?pr=134#step:6:21

Fixing src/scicookie/cookiecutter.json
Fixing src/scicookie/{{cookiecutter.project_slug}}/pyproject.toml

these issues should be caught by the pre-commit locally. check if you have it installed with `pre-commit install`

**Comment by xmnlab** [comment]
and run locally `pre-commit run --all-files --verbose`

File: src/scicookie/{{cookiecutter.project_slug}}/Makefile [hunk 3]
```diff
@@ -60,11 +60,17 @@ clean: ## remove build artifacts, compiled files, and cache
...
 	pre-commit run --all-files
 
-{% if cookiecutter.use_pytest %}
+{% if cookiecutter.use_pytest == "yes" -%}
 .PHONY:test
 test: ## run tests quickly with the default Python
 	pytest
-{% endif %}
+{%- endif +%}
+{% if cookiecutter.use_hypothesis == "yes" -%}
```
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/Makefile:68]: one problem here is that, 
if pytest == yes and hypothesis == yes, it will create the test target twice.

**Comment by ayeankit** [comment]
hey @xmnlab and @Saransh-cpp  again these test cases are failing , I am using pre-commit locally but here it not passing on gh workflows. I think I am not correctly dealing with makefile when pytest + hypothesis both options are selected. Can you look into it?

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+"""Tests for `{{ cookiecutter.package_slug }}` package."""
+{%- if cookiecutter.use_pytest == "yes" -%}
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:2]:
```suggestion
{%- if cookiecutter.use_pytest == "yes" %}
```

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import pytest
+{%- endif %}
+
+{%- if cookiecutter.use_hypothesis == "yes" -%}
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:6]:
```suggestion
{%- if cookiecutter.use_hypothesis == "yes" %}
```

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+from hypothesis import given, strategies as st
+{%- endif %}
+
+
+{%- if cookiecutter.use_pytest == "yes" -%}
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:11]:
```suggestion
{%- if cookiecutter.use_pytest == "yes" %}
```

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+@given(st.text())
+def test_content(response, hypothesis_argument):
+    """Sample pytest test function with the pytest fixture and hypothesis as arguments."""
+    # Test code using the response fixture and hypothesis_argument
+
+@given(st.text())
+def test_content(response, hypothesis_argument):
+    """Sample pytest test function with the hypothesis fixture as an argument."""
+    # Test code using the response fixture and hypothesis_argument
+{%- endif -%}
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:41]:
```suggestion
{% endif %}
{#- keep this line at the end of the file -#}
```

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+"""Tests for `{{ cookiecutter.package_slug }}` package."""
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:1]:
```suggestion
"""Tests for `{{ cookiecutter.package_slug }}` package.
"""
```

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    See more at: https://hypothesis.readthedocs.io/en/latest/quickstart.html
+    """
+{%- endif -%}
+
+
+{%- if cookiecutter.use_pytest == "yes" and cookiecutter.use_hypothesis == "yes" -%}
+@given(st.text())
+def test_content(response, hypothesis_argument):
+    """Sample pytest test function with the pytest fixture and hypothesis as arguments."""
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:34]:
```suggestion
    """Sample pytest test function with the pytest fixture and hypothesis as arguments.
    """
```

File: src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    # Test code using the response fixture and hypothesis_argument
+
+@given(st.text())
+def test_content(response, hypothesis_argument):
+    """Sample pytest test function with the hypothesis fixture as an argument."""
```
Comment (Anavelyz) [inline src/scicookie/{{cookiecutter.project_slug}}/tests/test_{{cookiecutter.package_slug}}.py:39]:
```suggestion
    """Sample pytest test function with the hypothesis fixture as an argument.
    """
```

**Review by Anavelyz — COMMENTED** [review]
@ayeankit, it seems that the docstrings are giving problems, I left you some suggestions. I hope it works.

**Comment by ayeankit** [comment]
@Anavelyz after changing the auto-string, it's still failing the gh actions. Can you see why its failing?

**Comment by xmnlab** [comment]
I didn't check the CI error ..but if it is just about the docstring .. just keep it shorter.
as it is just a placeholder, it does't matter too much.

**Comment by ayeankit** [comment]
Hey @xmnlab and @Anavelyz  Can you please review my PR again? I shorten the auto strings and locally its passes all the smoke test cases. But I don't know why it's failing here. Please suggest some way to pass the gh actions.

**Review by Saransh-cpp — COMMENTED** [review]
Sorry for the delay @ayeankit. From the tests it looks like there are some formatting issues in your PR. Remember, you can read the [logs of the failing tests](https://github.com/osl-incubator/scicookie/actions/runs/5382650276/jobs/9768368633?pr=134) through GitHub and you can also run them locally for easier debugging.

**Comment by ayeankit** [comment]
Hey @xmnlab , Finally all test cases passed. Can you review it once again?

**Review by xmnlab — APPROVED** [review]
LGTM! thanks @ayeankit !

**Comment by xmnlab** [comment]
@ayeankit , I almost forgot, could you add some documentation about hypothesis please?

you can add it here https://github.com/osl-incubator/scicookie/blob/main/docs/guide.md

and 

here https://github.com/osl-incubator/scicookie/blob/main/src/scicookie/%7B%7Bcookiecutter.project_slug%7D%7D/docs/contributing.md

You can take a look into @Anavelyz 's PR to have an idea: https://github.com/osl-incubator/scicookie/pull/136

Some few paragraphs about Hypothesis, plus a link to the Hypothesis documentation page would be great.

Additionally, add a line here in the feature section : https://github.com/osl-incubator/scicookie/blob/main/README.md
Just mentioning that the template has support for pytest and hypothesis for testing

File: README.md [hunk 11]
```diff
@@ -19,16 +19,13 @@ for a Python package.
...
+  - Test library: pytest, Hypothesis 
   - Auto format code tool: blue and black
   - Initial integration with git
   - Support to conda (as base environment) and poetry as packaging and dependency management
   - Support to pre-commit
   - CI with github actions
   - Release workflow with semantic release and github actions
-  - Flexible build system selection: Choose between popular build systems like
-      [Poetry](https://python-poetry.org/) or [Flit](https://flit.pypa.io) based
-      on your preference.
```
Comment (Saransh-cpp) [inline README.md:31]: I think this shouldn't have been removed.

File: docs/guide.md [hunk 12]
```diff
@@ -29,3 +29,12 @@ most popular system compilers to suit your needs and preferences for developing
...
 https://github.com/osl-incubator/scicookie/issues/new/choose.
+ 
+## Test Library
+
+There are several test library options available to development of Python packages. SciCookie support the following:
+
+
+  -  [**Pytest**](https://docs.pytest.org/en/7.3.x/contents.html): is a powerful property-based testing library for Python. It generates test cases based on properties, uncovering bugs and corner cases. Learn more in the Hypothesis documentation.
+
+- [**Hypothesis**](https://hypothesis.readthedocs.io/): is a powerful property-based testing library for Python. It generates test cases based on properties, uncovering bugs and corner cases. Learn more in the Hypothesis documentation.
```
Comment (Saransh-cpp) [inline docs/guide.md:40]: Some formatting issues and repetition here 🙂

File: README.md [hunk 13]
```diff
@@ -19,16 +19,13 @@ for a Python package.
   - Allows package slug (use `_` instead of `-`)
   - Licenses supported: MIT, BSD 3 Clause, ISC License, Apache Software License 2.0, and GPL 3
   - Documentation engines: mkdocs, sphinx, jupyter-boook
-  - Test library: pytest
+  - Test library: pytest, Hypothesis
```
Comment (Saransh-cpp) [inline README.md:22]:
```suggestion
  - Test library: pytest, hypothesis 
```

**Comment by ayeankit** [comment]
@xmnlab I changed the files as suggested by @Saransh-cpp. Can you review it once again?

File: docs/guide.md [hunk 14]
```diff
@@ -29,3 +29,13 @@ most popular system compilers to suit your needs and preferences for developing
 Python packages. If you think we should add more options, you can submit your
 suggestion as a issue at
 https://github.com/osl-incubator/scicookie/issues/new/choose.
+ 
+## Test Library
+
+There are several test library options available to development of Python packages. SciCookie support the following:
+
```
Comment (xmnlab) [inline docs/guide.md:36]:
```suggestion
```

File: docs/guide.md [hunk 15]
```diff
@@ -29,3 +29,13 @@ most popular system compilers to suit your needs and preferences for developing
...
+
```
Comment (xmnlab) [inline docs/guide.md:37]:
```suggestion
```

**Review by xmnlab — CHANGES_REQUESTED** [review]
@ayeankit in general it looks good. thanks!
could you just improve a bit the documentation in the template side please?

you can get some ideas from here https://github.com/osl-incubator/scicookie/pull/136/files#diff-f1c71bca02a04da875655ec4b14b5bc1ed65d7c025276bd97efbf97bc464a3b1

In general, you should try to see the documentation from the perspective of a new developer. Maybe a new developer would not know what is pytest or hypothesis.

so maybe 1 or 2 paragraphs about that and a link to the official web/documentation page would be great.

other than that, it looks great :)

**Comment by ayeankit** [comment]
@xmnlab I changed the files as suggested . Can you review it once again?

File: docs/guide.md [hunk 16]
```diff
@@ -29,3 +29,11 @@ most popular system compilers to suit your needs and preferences for developing
...
 suggestion as a issue at
 https://github.com/osl-incubator/scicookie/issues/new/choose.
+ 
+## Test Library
+
+There are several test library options available to development of Python packages. SciCookie support the following:
+
+-  [**Pytest**](https://docs.pytest.org/en/7.3.x/contents.html): is a popular testing framework for Python. It simplifies the process of writing and running tests by providing a concise syntax and powerful features. With Pytest, you can automatically discover and collect test cases, use fixtures for test setup and resource management, and write test functions with assert statements to check expected outcomes. It offers various options for test execution, including running specific tests, parallel execution, and generating test reports. Pytest also has a thriving ecosystem of plugins that extend its capabilities, such as code coverage analysis and test parameterization. Overall, Pytest is widely adopted for its simplicity, flexibility, and community support, making it an effective tool for ensuring the quality and reliability of Python code.You can check documentation [here](https://docs.pytest.org/en/7.3.x/contents.html)
```
Comment (Saransh-cpp) [inline docs/guide.md:37]:
```suggestion
-  [**Pytest**](https://docs.pytest.org/en/): is a popular testing framework for Python. It simplifies the process of writing and running tests by providing a concise syntax and powerful features. With Pytest, you can automatically discover and collect test cases, use fixtures for test setup and resource management, and write test functions with assert statements to check expected outcomes. It offers various options for test execution, including running specific tests, parallel execution, and generating test reports. Pytest also has a thriving ecosystem of plugins that extend its capabilities, such as code coverage analysis and test parameterization. Overall, Pytest is widely adopted for its simplicity, flexibility, and community support, making it an effective tool for ensuring the quality and reliability of Python code. You can check documentation [here](https://docs.pytest.org/en/)
```

File: docs/guide.md [hunk 17]
```diff
@@ -29,3 +29,11 @@ most popular system compilers to suit your needs and preferences for developing
...
+
+- [**Hypothesis**](https://hypothesis.readthedocs.io/): is a property-based testing library for Python. It focuses on generating diverse input data and exploring different scenarios to thoroughly test code. Instead of relying on specific examples, Hypothesis allows you to define general properties that your code should satisfy. It automatically generates random inputs, including edge cases, to uncover potential bugs and unexpected behaviors. Hypothesis integrates well with popular testing frameworks like Pytest and promotes comprehensive testing to improve code reliability.You can check documentation [here](https://hypothesis.readthedocs.io/)
```
Comment (Saransh-cpp) [inline docs/guide.md:39]:
```suggestion
- [**Hypothesis**](https://hypothesis.readthedocs.io/): is a property-based testing library for Python. It focuses on generating diverse input data and exploring different scenarios to thoroughly test code. Instead of relying on specific examples, Hypothesis allows you to define general properties that your code should satisfy. It automatically generates random inputs, including edge cases, to uncover potential bugs and unexpected behaviors. Hypothesis integrates well with popular testing frameworks like Pytest and promotes comprehensive testing to improve code reliability. You can check documentation [here](https://hypothesis.readthedocs.io/)
```

File: README.md [hunk 18]
```diff
@@ -19,16 +19,16 @@ for a Python package.
...
   - Auto format code tool: blue and black
   - Initial integration with git
   - Support to conda (as base environment) and poetry as packaging and dependency management
   - Support to pre-commit
   - CI with github actions
-  - Release workflow with semantic release and github actions
   - Flexible build system selection: Choose between popular build systems like
       [Poetry](https://python-poetry.org/) or [Flit](https://flit.pypa.io) based
       on your preference.
+  - Release workflow with semantic release and github actions
```
Comment (Saransh-cpp) [inline README.md:31]: This looks unrelated to this PR 🙂 

Should be okay though, just moved down

**Review by xmnlab — APPROVED** [review]
thanks for working on that @ayeankit !

**Comment by xmnlab** [comment]
@ayeankit next time, please provide more information in the `Pull Request description` section in the first comment in the PR. thanks
